namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala

```diff
@@ -212,27 +279,32 @@ class ThriftNamerInterface(
 
   private[this] val addrObserverCacheMu = new {}
   private[this] var addrObserverCache: Map[Path, AddrObserver] = Map.empty
-  private[this] def cacheBoundAddrs(nt: NameTree[Name.Bound]): Unit = nt match {
-    case NameTree.Neg =>
-    case NameTree.Empty =>
-    case NameTree.Fail =>
-    case NameTree.Alt(trees@_*) =>
-      trees.foreach(cacheBoundAddrs)
-    case NameTree.Union(trees@_*) =>
-      trees.foreach(w => cacheBoundAddrs(w.tree))
-    case NameTree.Leaf(bound) =>
-      bound.id match {
-        case id: Path =>
-          addrObserverCacheMu.synchronized {
-            if (!addrObserverCache.contains(id)) {
-              addrObserverCache += (id -> AddrObserver(bound.addr, stamper))
-            }
+  private[this] def observeAddr(ns: String, id: Path): AddrObserver =
+    addrObserverCacheMu.synchronized {
+      addrObserverCache.get(id) match {
+        case Some(obs) =>
+          Trace.recordBinary("namerd.srv/addr.cached", true)
+          obs
+        case None =>
+          Trace.recordBinary("namerd.srv/addr.cached", false)
+          val addr = getBind(ns, Dtab.empty, id).run.flatMap {
+            case Activity.Failed(e) => Var.value(Resolution.Resolved(Addr.Failed(e)))
+            case Activity.Pending => Var.value(Resolution.Resolved(Addr.Pending))
+            case Activity.Ok(NameTree.Leaf(bound)) if bound.id == id =>
+              bound.addr.map(Resolution.Resolved(_))
+            case _ => Var.value(Resolution.Released)
```

Comment:
This raises an interesting point: are there situations when Activity.Ok(tree) can be satisfied where NameTree is neither Leaf nor Neg? I think it's uncommon but certainly possible given the way this currently works.  And I don't think we can expect users to know to avoid these situations.

For example, given a dtab like:
```
/http/1.1/* => /io.l5d.k8s/prod/http ;
```
a bound name might be `/io.l5d.k8s/prod/http/web`.

What happens if we update the dtab so it looks like:
```
/http/1.1/* => /io.l5d.k8s/prod/http ;
/io.l5d.k8s/prod => /io.l5d.k8s/pre-prod ;
```

Now `/io.l5d.k8s/prod/http/web` binds to an Alt: `/io.l5d.k8s/pre-prod/http/web | /io.l5d.k8s/prod/http/web`? (Is that right? Or does that go infinitely recursive?)

Or more simply, we may find something like:
```
/http/1.1/* => /io.l5d.k8s/prod/http ;
/io.l5d.k8s/prod/http/web => /$/empty ;
```

Perhaps we can avoid this by doing addr lookups directly against a namer instead of through via a dtab?

Or perhaps we can simply forbid dtabs that contain namer prefixes on the LHS? This seems like a generally good policy in any regard.